### ISSUE TITLE
Build failure with Conduit

### ISSUE THREAD

**Issue by JaeseungYeom** [LLNL/lbann#566]
When compiled with '--with-conduit' option, I am getting a linking error as below:
```
--------------------------------------------------------------------------------------
[ 84%] Built target lbann
[ 84%] Linking CXX executable jag_converter
../../liblbann.so: undefined reference to `lbann::jag_store::load_images(std::vector<std::string, std::allocator<std::string> > const&)'
../../liblbann.so: undefined reference to `lbann::jag_store::jag_store()'
../../liblbann.so: undefined reference to `lbann::jag_store::load_inputs()'
../../liblbann.so: undefined reference to `lbann::jag_store::setup(std::vector<std::string, std::allocator<std::string> >, bool, int)'
../../liblbann.so: undefined reference to `lbann::jag_store::get_linearized_data_size() const'
collect2: error: ld returned 1 exit status
make[2]: *** [src/data_store/CMakeFiles/jag_converter-bin.dir/build.make:164: src/data_store/jag_converter] Error 1
make[1]: *** [CMakeFiles/Makefile2:915: src/data_store/CMakeFiles/jag_converter-bin.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
[  1%] Built target LbannProto
[ 84%] Built target lbann
[ 84%] Linking CXX executable jag_converter
../../liblbann.so: undefined reference to `lbann::jag_store::load_images(std::vector<std::string, std::allocator<std::string> > const&)'
../../liblbann.so: undefined reference to `lbann::jag_store::jag_store()'
../../liblbann.so: undefined reference to `lbann::jag_store::load_inputs()'
../../liblbann.so: undefined reference to `lbann::jag_store::setup(std::vector<std::string, std::allocator<std::string> >, bool, int)'
../../liblbann.so: undefined reference to `lbann::jag_store::get_linearized_data_size() const'
collect2: error: ld returned 1 exit status
make[2]: *** [src/data_store/CMakeFiles/jag_converter-bin.dir/build.make:164: src/data_store/jag_converter] Error 1
make[1]: *** [CMakeFiles/Makefile2:915: src/data_store/CMakeFiles/jag_converter-bin.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
--------------------------------------------------------------------------------------
```
If I comment out data_reader_jag_conduit_hdf5 and jag_store from CMakeList, lbann.hpp, and init_image_data_reader.cpp, it compiles. However, it is not obvious to me why this error occurs.

**Comment by timmoon10** [comment]
Maybe the solution to #562 was a little too aggressive.

**Comment by JaeseungYeom** [comment]
I see the issue there.  `#ifdef LBANN_HAS_CONDUIT` comes before #include `"lbann_config.hpp"`. I will do some testing and check in the fix.

And I just updated the code and I see new error on sierra as below (using build_lbann_lc.sh --with-conduit --clean-build):
```
CMake Warning:
  Manually-specified variables were not used by the project:

    LBANN_BUILT_WITH_SPECTRUM
    protobuf_BUILD_SHARED_LIBS


-- Build files have been written to: /p/gscratch1/yeom2/LBANN/lbann-org/build/gnu.Release.sierra.llnl.gov/lbann/build
make[2]: *** [lbann/CMakeFiles/LBANN.dir/build.make:112: lbann/stamp/LBANN-configure] Error 1
make[2]: Leaving directory '/p/gscratch1/yeom2/LBANN/lbann-org/build/gnu.Release.sierra.llnl.gov'
make[1]: *** [CMakeFiles/Makefile2:476: lbann/CMakeFiles/LBANN.dir/all] Error 2
make[1]: Leaving directory '/p/gscratch1/yeom2/LBANN/lbann-org/build/gnu.Release.sierra.llnl.gov'
make: *** [Makefile:84: all] Error 2
--------------------
MAKE FAILED
--------------------
```
when used --with-conduit -j 1 after that, it showed
```
nvcc fatal   : No input files specified; use option --help for more information
```

**Comment by JaeseungYeom** [comment]
The new error is gone now thanks to the last update. The original issue still remains. placing the header inclusion before the macro check both in jag_store.hpp and jag_store.cpp did not fully resolved the issue.
The current error I am seeing is
```
[100%] Linking CXX executable lbann
cd /p/lscratchh/yeom2/LBANN1/lbann/build/gnu.Release.catalyst.llnl.gov/lbann/build/model_zoo && /usr/tce/packages/cmake/cmake-3.9.2/bin/cmake -E cmake_link_script CMakeFiles/lbann-bin.dir/link.txt --verbose=0
../liblbann.so: undefined reference to `conduit::relay::io::hdf5_group_list_child_names(int, std::string const&, std::vector<std::string, std::allocator<std::string> >&)'
../liblbann.so: undefined reference to `conduit::relay::io::hdf5_close_file(int)'
../liblbann.so: undefined reference to `conduit::relay::io::hdf5_read(int, std::string const&, conduit::Node&)'
```
However, gnu.Release.XXX.llnl.gov/lbann/build/liblbann.so seems to be linked with the library.
The undefined references above can be found in libconduit_relay.so
```
        libconduit.so => /p/lscratchh/brainusr/conduit/install-toss3-dev/lib/libconduit.so (0x00002aaab05ff000)
        libconduit_relay.so => /p/lscratchh/brainusr/conduit/install-toss3-dev/lib/libconduit_relay.so (0x00002aaab090a000)
        libconduit_blueprint.so => /usr/workspace/wsb/icfsi/conduit/install-toss3-dev/lib/libconduit_blueprint.so (0x00002aaab29e5000)
        libhdf5.so.1000 => /usr/workspace/wsb/icfsi/conduit-helpers/hdf5-quartz/hdf5/lib/libhdf5.so.1000 (0x00002aaab2c78000)
        libsz.so.2 => /usr/workspace/wsb/icfsi/conduit-helpers/szip-quartz/szip/lib/libsz.so.2 (0x00002aaab35ef000)
```

**Comment by JaeseungYeom** [comment]
addressed by PR #575 and #576. I will close when these two get merged, and I will merge them in a couple of hour if not merged by then.
